eclipse-theia/theia
# fix: various debug fixes and VS Code compatibility enhancements

#### What it does


 - feat: added support for `debug/toolbar` and `debug/variables/context`,
 - feat: added support for `debugState` when context (https://github.com/eclipse-theia/theia/issues/11871),
 - feat: can customize debug session timeout, and error handling (https://github.com/eclipse-theia/theia/issues/11879),
 - fix: the `debugType` context that was not updated on session stop,
 - fix: `configure` must happen after receiving capabilities (https://github.com/eclipse-theia/theia/issues/11886),
 - fix: added missing conext menu in the _Variables_ view,
 - fix: handle `setFunctionBreakboints` response with no `body` (https://github.com/eclipse-theia/theia/issues/11885),
 - fix: `DebugExt` fires `didStart` event on `didCreate` (https://github.com/eclipse-theia/theia/issues/11916),
 - fix: validate editor selection based on the text model (https://github.com/eclipse-theia/theia/issues/11880)

Closes #11871
Closes #11879
Closes #11885
Closes #11886
Closes #11916
Closes #11880

#### How to test


 - Get the modified version of the `vscode.mock-debug` extension from my fork: https://github.com/kittaakos/vscode-mock-debug/blob/theia/mock-debug-0.51.0.vsix. Here is the diff with the changes from the upstream version: https://github.com/microsoft/vscode-mock-debug/compare/main...kittaakos:vscode-mock-debug:theia.
 - Open the [`./vscode-mock-debug/sampleWorkspace`](https://github.com/kittaakos/vscode-mock-debug/tree/theia/sampleWorkspace) folder as a workspace in Theia. There is a readme file in it and the `launch.json` in the `.vscode` folder.
 - Press <kbd>F5</kbd>, `readme.md` opens, and the top stack frame is the first line of the readme file. Check the `DevTools` console; there is no `Error: Illegal value for lineNumber`. ([#11880](https://github.com/eclipse-theia/theia/issues/11880))

https://user-images.githubusercontent.com/1405703/207561043-1194968d-5855-40a0-b257-e6d5cae56569.mp4

 - Set a breakpoint before any variable declarations (line 17) and after (line 28). You can use other lines too. Continue the debugger. It hits the breakpoint. ([#11885](https://github.com/eclipse-theia/theia/issues/11885)). You see no local variable. Continue the debugger. It hits the next breakpoint when there are variables. It's visible from the _Variables_ view. Right-click the variable, and you will see the `debugType` and `debugState` dependent content menu from the VSIX. Click on it. It works. You also see two additional debug toolbar items—one with the ❤️ `codicon` and the other with text. Click on them, and both work. ([#11871](https://github.com/eclipse-theia/theia/issues/11871))

https://user-images.githubusercontent.com/1405703/207563422-7a496e8b-521b-4246-ab06-1a2912f47b9a.mp4

#### For reviewers:

> * fix: `DebugExt` fires `didStart` event on `didCreate` ([[debug] debug-ext prematurely fires `didStart` event on `didCreate` #11916](https://github.com/eclipse-theia/theia/issues/11916)),

This 👆  requires a more complex setup and another VSIX (`marus25.cortex-debug-1.5.1.vsix` from https://downloads.arduino.cc/marus25.cortex-debug/marus25.cortex-debug-1.5.1.vsix). The fix cannot be verified with the `mock-debug` extension. Also, I used a physical board (Arduino Zero) where the code runs. The code runs on the device, and the debugger talks to it via dgb. I am happy to provide the steps to set up the toolchain, but the board is needed. Let me know it's required.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
